lib/ABI/FunctionType/Layout.cpp

```diff
@@ -657,6 +657,43 @@ uint64_t finalStackOffset(const model::CABIFunctionType &Function) {
   return ToRawConverter(ABI).finalStackOffset(CombinedStackArgumentSize);
 }
 
+UsedRegisters usedRegisters(const model::CABIFunctionType &Function) {
```

Comment:
I see a quite a bit of newly added code, I was hoping that this changeset would just outline pieces of things that were buried in `Layout` and the other classes...

Reply:
I tried that edition first. But that made layout itself a bit more complex.

Consider that practically all the complexity here is because we want to handle SPTAR gracefully.

Without it, it's just
```
UsedRegisters usedRegisters(const model::CABIFunctionType &Function) {
  auto Result.RVs = move_return_value_registers_in();
  
  handle_SPTAR_gracefully();
  
  for (const auto *Argument = Function.Arguments().tryGet(Index++);
     Argument != nullptr;
     Argument = Function.Arguments().tryGet(Index++)) {
     // (if it's this complicated looking loop you don't like, we could use some range-magic and hide it under something like `TrackingContainer::iterate`)
     // The only reason I didn't is because it's only ever usable for CFT-Arguments (because it's just a vector disguised as a `SortedVector`,
     // So, there's not many points where we could reuse it, AND most of those are already happy relying on `usedRegisters` as is - so here feels general enough.
             std::ranges::move(distributeArgument(...).Registers,
                          std::back_inserter(Result.Arguments));
  }
}
```

Reply:
OK, comment the function about what each stage is doing.

> this complicated looking loop you don't like...

Why can't you just iterate over the arguments? What situation are you trying to handle?
This `SortedVector` should have keys from `0` to its size `- 1`. If this is not checked in `::verify`, we should check it there and assume it here.

Reply:
1. Done.

2. We don't need to read _every_ argument in every case. For some ABIs even as little as one argument is enough, hence why it's easier to use index - this lets to not call `end` and otherwise rely on the number of arguments in any way.

Reply:
2. Ah, that's a pretty obscure feature to have without documentation. In any case, let's not push optimizations this far, unless we have proof that's beneficial. We're reading one argument, I'm OK with reading them all. Let's favor simplicity.

Reply:
I went along and "hid" the complexity within the `TrackingContainer`.
This function now looks like
```cpp
UsedRegisters usedRegisters(const model::CABIFunctionType &Function) {
  UsedRegisters Result;

  // Ready the return value register data.
  const abi::Definition &ABI = abi::Definition::get(Function.ABI());
  auto RV = ToRawConverter(ABI).distributeReturnValue(Function.ReturnType());
  std::ranges::move(RV.Registers, std::back_inserter(Result.ReturnValues));

  // Handle shadow pointer return value gracefully.
  ArgumentDistributor Distributor(ABI);
  if (RV.SizeOnStack != 0) {
    Distributor.addShadowPointerReturnValueLocationArgument();

    revng_assert(Result.ReturnValues.empty());
    const auto &GPRs = ABI.GeneralPurposeReturnValueRegisters();
    revng_assert(!GPRs.empty());
    Result.ReturnValues.emplace_back(GPRs[0]);

    if (ABI.ReturnValueLocationRegister() != model::Register::Invalid)
      Result.Arguments.emplace_back(ABI.ReturnValueLocationRegister());
  }

  // Iterate over arguments until we sure no further argument can use
  // any registers.
  for (const model::Argument &Argument : Function.Arguments().asVector()) {
    auto Distributed = Distributor.nextArgument(Argument.Type());
    for (DistributedValue SingleEntry : Distributed)
      if (!SingleEntry.RepresentsPadding)
        std::ranges::move(SingleEntry.Registers,
                          std::back_inserter(Result.Arguments));

    if (!Distributor.canNextArgumentUseRegisters())
      break;
  }

  return Result;
}
```